java/gradle.java/test/unit/src/org/netbeans/modules/gradle/queries/GradleSourceAttacherImplTest.java

```diff
@@ -0,0 +1,176 @@
+                        + "repositories {\n" +
+                        "    mavenCentral()\n"
+                       + "}\n"
+                       + "\n"
+                       + "dependencies {\n"
+                       + "    implementation(\"org.netbeans.api:org-openide-util:RELEASE124\")\n"
+                       + "}"
+               );
+        FileObject src = FileUtil.createFolder(fo, "src/main/java");
+        FileObject java = src.createData("Whatever.java");
+        Project prj = ProjectManager.getDefault().findProject(fo);
+        assertNotNull(prj);
+        ProjectTrust.getDefault().trustProject(prj);
+
+        OpenProjects.getDefault().open(new Project[] { prj }, false);
+        Project[] opened = OpenProjects.getDefault().openProjects().get();
+        
+        ActionProvider ap = prj.getLookup().lookup(ActionProvider.class);
+        
+        
+        class AP extends ActionProgress {
+            final CountDownLatch l = new CountDownLatch(1);
+            volatile boolean success;
+            
+            @Override
+            protected void started() {
+            }
+
+            @Override
+            public void finished(boolean success) {
+                this.success = success;
+                l.countDown();
+            }
+        }
+        
+        
+        AP prg1 = new AP();
+        
+        ap.invokeAction(ActionProvider.COMMAND_BUILD, Lookups.singleton(prg1));
+        prg1.l.await();
+        
+        assertTrue("Build, including jar download succeeded - requires maven central access", prg1.success);
+
+        GradleJavaProject gp = GradleJavaProject.get(prj);
+        GradleBaseProject gbp = GradleBaseProject.get(prj);
+        GradleConfiguration cfg = gbp.getConfigurations().get("compileClasspath");
+        Set<ModuleDependency> dep = cfg.findModules("org.netbeans.api:org-openide-util:RELEASE124");
+        assertNotNull(dep);
+        assertEquals(1, dep.size());
+        Set<File> files = dep.iterator().next().getArtifacts();
+        assertNotNull(files);
+        assertEquals(1, files.size());
+        
+        File utilJar = files.iterator().next();
+        
+        // now check that the javadoc attacher is applied:
+        class L implements SourceJavadocAttacher.AttachmentListener {
+            CompletableFuture<Boolean> result = new CompletableFuture<>();
+            
+            @Override
+            public void attachmentSucceeded() {
+                result.complete(true);
+            }
+
+            @Override
+            public void attachmentFailed() {
+                result.complete(false);
+            }
+        }
+        
+        // need to get the classpath root inside the archive.
+        URL archiveRoot = FileUtil.getArchiveRoot(utilJar.toURI().toURL());
+        L l = new L();
+        SourceJavadocAttacher.attachSources(archiveRoot, l);
+        // wait at most 20 seconds for a download.
+        boolean r = l.result.get(20, TimeUnit.SECONDS);
```

Comment:
I suspect this might lead to random failures, if the artifacts cannot be downloaded quickly enough. Possibly not too problematic, but thought I'd point that out.
Maybe it would be possible to setup a repository with dummy jars on the local disk and the do something like:
```
    maven {
        url "file://<path>"
    }
```
instead of `mavenCentral()` in `build.gradle`?

Reply:
Hm ! Certainly that would improve the stability, especially in the case of a clean gradle user home.